dysnix/app/templates/_helpers.tpl

```diff
@@ -1,56 +1,132 @@
 {{/* vim: set filetype=mustache: */}}
 
 {{/*
-Return the proper App image name
+  Returns the chart operation mode direct or library
 */}}
-{{- define "app.image" -}}
-{{ include "common.images.image" (dict "imageRoot" .Values.image "global" .Values.global) }}
+{{- define "app.chart.mode" -}}
+  {{- ternary "direct" "library" (eq .Chart.Name "app") -}}
+{{- end -}}
+
+{{/*
+  Returns proper chart name based on the operation mode
+*/}}
+{{- define "app.chart.name" -}}
+  {{- ternary ._include.topValues.app.name .Chart.Name (eq "direct" (include "app.chart.mode" .)) -}}
+{{- end -}}
+
+{{/*
+  Component label
+*/}}
+{{- define "app.labels.component" -}}
+  {{- if ._include.component -}}
+    app.kubernetes.io/component: {{ ._include.component }}
+  {{- end -}}
 {{- end -}}
 
 {{/*
 Create the name of the service account to use
 */}}
 {{- define "app.serviceAccountName" -}}
-{{- if .Values.serviceAccount.create -}}
-    {{ default (include "common.names.fullname" .) .Values.serviceAccount.name }}
-{{- else -}}
-    {{ default "default" .Values.serviceAccount.name }}
+  {{- $sa := .Values.serviceAccount | default dict -}}
+  {{- if  $sa.create -}}
+    {{- default (include "common.names.fullname" .) $sa.name -}}
+  {{- else if and ._include.topValues.serviceAccount.create $sa.reuse -}}
+    {{- default (include "common.names.fullname" .) ._include.topValues.serviceAccount.name -}}
+  {{- else -}}
+    {{- default "default" $sa.name -}}
+  {{- end -}}
+{{- end -}}
+
+{{/*
+Return the proper image name (for the init container volume-permissions image)
+*/}}
+{{- define "app.volumePermissions.image" -}}
+  {{- include "common.images.image" (dict "imageRoot" .Values.volumePermissions.image "global" .Values.global) -}}
 {{- end -}}
+
+{{/*
+Return service ports
+*/}}
+{{- define "app.service.ports" -}}
+  {{- $ports := list -}}
+  {{- $yaml := printf "ports:\n%s" (include "app.tplvalues.named-list" (dict "value" .Values.service.ports "valueKey" "port" "context" $)) -}}
+
+  {{- range $_, $val := get ($yaml | fromYaml) "ports" -}}
+    {{/* Set nodePort to null for type ClusterIP */}}
```

Comment:
Why do we need to set nodePort to null with ClusterIP ?

Reply:
Let's elaborate this. For me looks like for ClusterIP service type nodePort is not valid... And this looks like a switch helper which removes the nodePort (on a transition from NodePort to ClusterIP type). 

_Since ClusterIP and NodePort values are immutable – they cannot be changed throughout the lifecycle of an object. The only way to change or update an existing Service is to provide the right metadata and omit both ClusterIP and NodePort values from the spec._

Reply:
>Since ClusterIP and NodePort values are immutable

Hm, I've changed `.spec.type` from `ClusterIP` to `NodePort` and back w/o issues via `kubectl edit` on `v1.22.15-gke.2500`. 
Indeed, there is no option to change `.spec.clusterIP` value itself. But `.spec.ports[*].nodePort` value is changeable in my tests.
Also there is no need to remove `.spec.ports[*].nodePort` while altering `.spec.type` from `NodePort` to `ClusterIP`. Are you aware of any real life cases where this feature will be required/useful?

Reply:
Thanks for testing, I will drop this code then.